File: bgm-code/ParameterEstimation_Stat/ExamplesBinomial_Estimation/Rate_2.m
% Difference Between Two Rates

clear;

% Data (Observed Variables)
k1=5;n1=10;k2=7;n2=10

% WinBUGS Parameters
nchains=1; % How Many Chains?
nburnin=0; % How Many Burn-in Samples?
nsamples=1e4; % How Many Recorded Samples?

% Assign Matlab Variables to the Observed WinBUGS Nodes
datastruct = struct('k1',k1,'n1',n1,'k2',k2,'n2',n2);

% Initialize Unobserved Variables
for i=1:nchains
    S.theta1 = 0.5; % An Intial Value for the Success Rate
    S.theta2 = 0.5; % An Intial Value for the Success Rate
    init0(i) = S;
end

% Use WinBUGS to Sample
[samples, stats] = matbugs(datastruct, ...
    fullfile(pwd, 'Rate_2.txt'), ...
    'init', init0, ...
    'nChains', nchains, ...
    'view', 1, 'nburnin', nburnin, 'nsamples', nsamples, ...
    'thin', 1, 'DICstatus', 0, 'refreshrate',100, ...
    'monitorParams', {'theta1','theta2','delta'}, ...
    'Bugdir', 'C:/Program Files/WinBUGS14');

% Analysis
figure(1);clf;hold on;
eps=.025;bins=[-1:eps:1];
count=hist(samples.delta,bins);
count=count/sum(count)/eps;
ph=plot(bins,count,'k-');
set(gca,'box','on','fontsize',14,'xtick',[-1:.2:1]);
xlabel('Difference in Rates','fontsize',16);
ylabel('Posterior Density','fontsize',16);

% Summaries of Posterior
% MEAN
disp(sprintf('Mean is %1.2f',mean(samples.delta)));
% MODE
logL=zeros(nsamples,1);
for i=1:nsamples
	logL(i)=logL(i)+k1*log(samples.theta1(i))+(n1-k1)*log(1-samples.theta1(i));
	logL(i)=logL(i)+k2*log(samples.theta2(i))+(n2-k2)*log(1-samples.theta2(i));
end;
[val ind]=max(logL);
disp(sprintf('Mode is %1.2f',samples.delta(ind)));
% MEDIAN
disp(sprintf('Median is %1.2f',median(samples.delta)));
% CREDIBLE INTERVAL
cred=0.95;
b1=(1-cred)/2;b2=1-b1;
val=sort(samples.delta);
disp(sprintf('%d percent credible interval is [%1.2f, %1.2f]',cred*100,val(round(b1*nsamples)),val(round(b2*nsamples))));

File: bgm-code/ParameterEstimation_Stat/ExamplesBinomial_Estimation/Rate_2.R
# clears workspace:  
rm(list=ls(all=TRUE)) 

# sets working directories:
#setwd("F:/Codes/ExamplesBinomial_Estimation")
setwd("C:/Documents and Settings/ewagenmakers/My Documents/My Dropbox/EJ/temp/BayesBook/Code/ParameterEstimation_Stat/ExamplesBinomial_Estimation")

library(R2WinBUGS)
bugsdir = "C:/Program Files/WinBUGS14"

k1 = 5
k2 = 7
n1 = 10
n2 = 10

data  = list("k1", "k2", "n1", "n2") # to be passed on to WinBUGS
myinits =	list(
  list(theta1 = 0.1, theta2 = 0.9))

# parameters to be monitored:	
parameters = c("delta", "theta1", "theta2")

# The following command calls WinBUGS with specific options.
# For a detailed description see Sturtz, Ligges, & Gelman (2005).
samples = bugs(data, inits=myinits, parameters,
	 			model.file ="Rate_2.txt",
	 			n.chains=1, n.iter=10000, n.burnin=0, n.thin=1,
	 			DIC=T, bugs.directory=bugsdir,
	 			codaPkg=F, debug=F)
# Now the values for the monitored parameters are in the "samples" object, 
# ready for inspection.

# Collect posterior samples:
delta = samples$sims.list$delta 

# Plot a density estimate for delta: 
plot(density(delta), xlab="Difference in Rates", ylab="Posterior Density", main =" ", xlim=c(-1,1))

# Add some plotting options to make things look better:
par(cex.main = 1.5, mar = c(5, 6, 4, 5) + 0.1, mgp = c(3.5, 1, 0), cex.lab = 1.5,
    font.lab = 2, cex.axis = 1.3, bty = "n", las=1)
plot(density(delta), xlab="Difference in Rates", ylab="Posterior Density", main =" ", lwd=2, xlim=c(-1,1))
 
# mean of delta:
mean(delta)
# median of delta:
median(delta)
# mode of delta, estimated from the "density" smoother:
density(delta)$x[which(density(delta)$y==max(density(delta)$y))]
# 95% confidence interval for delta:
low  = round(.025*length(delta))
high = round(.975*length(delta))
sort(delta)[low]
sort(delta)[high]
